core/state/statedb.go

```diff
@@ -502,12 +494,17 @@ func (s *StateDB) updateStateObject(obj *stateObject) {
 		defer func(start time.Time) { s.AccountUpdates += time.Since(start) }(time.Now())
 	}
 	// Encode the account and update the account trie
-	addr := obj.Address()
-	if err := s.trie.UpdateAccount(addr, &obj.data); err != nil {
-		s.setError(fmt.Errorf("updateStateObject (%x) error: %v", addr[:], err))
+	tr, err := s.accountTrie()
+	if err != nil {
+		s.setError(fmt.Errorf("failed to load account trie (%x) error: %v", obj.Address(), err))
+		return
+	}
+	if err := tr.UpdateAccount(obj.Address(), &obj.data); err != nil {
+		s.setError(fmt.Errorf("updateStateObject (%x) error: %v", obj.Address(), err))
+		return
 	}
```

Comment:
Actually, I stop the abstraction for StateWriter because of high complexity.

For a stateDB with "Writing && Hashing" functionalities, a few elements are mandatory, including the `Trie` for computing the new state root. Because Trie is required in any case, I keep this notion to avoid unnecessary over-abstraction situation.

The components we can hide is `Snapshot`, which is optional. We can abstract it a bit, but it's not relevant with Trie at all.

Reply:
but is `Trie` really necessary in any case? I thought that you wanted to be able to work snapshot-only as well?

In the verkle case, for sure the account trie is useful and I'm not complaining if it's enshrined. What greatly bothers me, and I believe should be abstracted, is the storage trie.

Reply:
Regarding the writing/hashing part, I also hope that this can be abstracted, although I understand this would be too big of a change to be added to this PR. I just want to confirm that this is the goal, because if we only have a state reader abstraction without a state hasher or state writer, I don't think a standalone reader abstraction makes a lot of sense (for my use case, that is)

Reply:
> but is Trie really necessary in any case?

Trie is not required in case of a "Non-Writable" stateDB, specifically we only read states and modify states, but we never commit/rehash them. Otherwise, trie is mandatory.

> and I believe should be abstracted, is the storage trie.

I understand the pain, but no matter how we abstract the thing, we still need to read the account trie and storage tries from the database, cache them within the stateDb. In case of verkle, we use the single trie for both account trie and storage tries, it's the only difference. But yeah, I agree the existent if/else control for verkle mode is ugly.

> Regarding the writing/hashing part

The existent abstraction happens in the Trie itself. We use different rules for writing/hashing mechanism. My point is `StateHasher` or `StateWrite` is doable, but maybe some things over abstracted.